### PR TITLE
Made the app_port property in the dapr block of the azurerm_container_app resource optional

### LINKED ISSUE #20534: The app_port property in the dapr block of the azurerm_container_app resource should be optional and not required
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

 v1.3.7

### AzureRM Provider Version

v3.43.0

### Affected Resource(s)/Data Source(s)

azurerm_container_app 

### Terraform Configuration Files

```hcl
variable "container_apps" {
  description = "Specifies the container apps in the managed environment."
  type = list(object({
    name                           = string
    revision_mode                  = optional(string)
    ingress                        = optional(object({
      allow_insecure_connections   = optional(bool)
      external_enabled             = optional(bool)
      target_port                  = optional(number)
      transport                    = optional(string)
      traffic_weight               = optional(list(object({
        label                      = optional(string)
        latest_revision            = optional(bool)
        revision_suffix            = optional(string)
        percentage                 = optional(number)
      })))
    }))
    dapr                           = optional(object({
      app_id                       = optional(string)
      app_port                     = optional(number)
      app_protocol                 = optional(string)
    }))
    secrets                        = optional(list(object({
      name                         = string
      value                        = string
    })))
    template                       = object({
      containers                   = list(object({
        name                       = string
        image                      = string
        args                       = optional(list(string))
        command                    = optional(list(string))
        cpu                        = optional(number)
        memory                     = optional(string)
        env                        = optional(list(object({
          name                     = string
          secret_name              = optional(string)
          value                    = optional(string)
        })))
      }))
      min_replicas                 = optional(number)
      max_replicas                 = optional(number)
      revision_suffix              = optional(string)
      volume                       = optional(list(object({
        name                       = string
        storage_name               = optional(string)
        storage_type               = optional(string)
      })))
    })
  }))
  default                          = [{
    name                           = "nodeapp"
    revision_mode                  = "Single"
    ingress                        = {
      external_enabled             = false
      target_port                  = 3000
      transport                    = "http"
      traffic_weight               = [{
        label                      = "blue"
        latest_revision            = true
        revision_suffix            = "blue"
        percentage                 = 100
      }]
    }
    dapr                           = {
      app_id                       = "nodeapp"
      app_port                     = 3000
      app_protocol                 = "http"
    }
    template                       = {
      containers                   = [{
        name                       = "hello-k8s-node"
        image                      = "dapriosamples/hello-k8s-node:latest"
        cpu                        = 0.5
        memory                     = "1Gi"
        env                        = [{
          name                     = "APP_PORT"
          value                    = 3000
        }]
      }]
      min_replicas                 = 1
      max_replicas                 = 1
    }
  },
  {
    name                           = "pythonapp"
    revision_mode                  = "Single"
    dapr                           = {
      app_id                       = "pythonapp"
      app_port                     = 80
    }
    template                       = {
      containers                   = [{
        name                       = "hello-k8s-python"
        image                      = "dapriosamples/hello-k8s-python:latest"
        cpu                        = 0.5
        memory                     = "1Gi"
      }]
      min_replicas                 = 1
      max_replicas                 = 1
    }
  }]
}

resource "azurerm_container_app" "container_app" {
  for_each                     = {for app in var.container_apps: app.name => app}

  name                         = each.key
  resource_group_name          = var.resource_group_name
  container_app_environment_id = azurerm_container_app_environment.managed_environment.id
  tags                         = var.tags
  revision_mode                = each.value.revision_mode

  template {
    dynamic "container" {
      for_each                   = coalesce(each.value.template.containers, [])
      content {
        name                     = container.value.name
        image                    = container.value.image
        args                     = try(container.value.args, null)
        command                  = try(container.value.command, null)
        cpu                      = container.value.cpu
        memory                   = container.value.memory
        
        dynamic "env" {
          for_each               = coalesce(container.value.env, [])
          content {
            name                 = env.value.name
            secret_name          = try(env.value.secret_name, null)
            value                = try(env.value.value, null)
          }
        }
      }
    }
    min_replicas                 = try(each.value.template.min_replicas, null)
    max_replicas                 = try(each.value.template.max_replicas, null)
    revision_suffix              = try(each.value.template.revision_suffix, null)

    dynamic "volume" {
      for_each                   = each.value.template.volume != null ? [each.value.template.volume] : []
      content {
        name                     = volume.value.name
        storage_name             = try(volume.value.storage_name, null)
        storage_type             = try(volume.value.storage_type, null)
      }
    }
  }

 dynamic "ingress" {
    for_each                     = each.value.ingress != null ? [each.value.ingress] : []
    content {
      allow_insecure_connections = try(ingress.value.allow_insecure_connections, null)
      external_enabled           = try(ingress.value.external_enabled, null)
      target_port                = ingress.value.target_port
      transport                  = ingress.value.transport

      dynamic "traffic_weight"  {
        for_each                 = coalesce(ingress.value.traffic_weight, [])
        content {
          label                  = traffic_weight.value.label
          latest_revision        = traffic_weight.value.latest_revision
          revision_suffix        = traffic_weight.value.revision_suffix
          percentage             = traffic_weight.value.percentage
        }
      }
    }
  }

  dynamic "dapr" {
    for_each                     = each.value.dapr != null ? [each.value.dapr] : []
    content {
      app_id                     = dapr.value.app_id
      app_port                   = dapr.value.app_port
      app_protocol               = dapr.value.app_protocol
    }
  }

  dynamic "secret" {
    for_each                     = each.value.secrets != null ? [each.value.secrets] : []
    content {
      name                       = secret.value.name
      value                      = secret.value.value
    }
  }

  lifecycle {
    ignore_changes = [
        tags
    ]
  }
}
```

### Debug Output/Panic Output

```shell
The resource is successfully created:


module.container_apps.azurerm_container_app.container_app["pythonapp"]: Creation complete after 39s [id=/subscriptions/1a45a694-ae23-4650-9774-89a571c462f6/resourceGroups/AnubisRG/providers/Microsoft.App/containerApps/pythonapp]
```

### Expected Behaviour

In the `azurerm_container_app` resource, the `app_port` property under the `dapr` block should be optional and not required. You should be able to set the value of this property to `null` to create headless applications, like the `pythonapp` in the [Tutorial: Deploy a Dapr application to Azure Container Apps with an Azure Resource Manager or Bicep template](https://learn.microsoft.com/en-us/azure/container-apps/microservices-dapr-azure-resource-manager?tabs=bash&pivots=container-apps-bicep), with no ingress, hence with no `app_port`.

### Actual Behaviour

You are obliged to specify a value for the `app_port` property under the `dapr` block of an `azurerm_container_app` resource even if the application does not expose any ingress.must

### Steps to Reproduce

Just try to create a headless application, like the `pythonapp` in the [Tutorial: Deploy a Dapr application to Azure Container Apps with an Azure Resource Manager or Bicep template](https://learn.microsoft.com/en-us/azure/container-apps/microservices-dapr-azure-resource-manager?tabs=bash&pivots=container-apps-bicep), with no ingress, hence with no `app_port`.

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by paolosalvatori**
This pull  request makes the `app_port` property in the `dapr` block of the `azurerm_container_app` resource should optional instead of required. As explained in the related issue [#20534](https://github.com/hashicorp/terraform-provider-azurerm/issues/20534), you should be able to set the value of this property to `null` to create headless applications, like the `pythonapp` in the [Tutorial: Deploy a Dapr application to Azure Container Apps with an Azure Resource Manager or Bicep template](https://learn.microsoft.com/en-us/azure/container-apps/microservices-dapr-azure-resource-manager?tabs=bash&pivots=container-apps-bicep), with no ingress, hence with no `app_port`. This PR resolves #20534

### CODE REVIEW

**Comment by zioproto** [comment]
Cc: @lonegunmanb

File: internal/services/containerapps/helpers/container_apps.go [hunk 1]
```diff
@@ -404,7 +404,7 @@ func ContainerDaprSchema() *pluginsdk.Schema {
 
 				"app_port": {
 					Type:        pluginsdk.TypeInt,
-					Required:    true,
+					Required:    false,
```
Comment (lonegunmanb) [inline internal/services/containerapps/helpers/container_apps.go:407]: Better be `Optional: true`

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
A suggestion on your pr @paolosalvatori

File: internal/services/containerapps/helpers/container_apps.go (hunk 1 above)
Comment (paolosalvatori) [reply internal/services/containerapps/helpers/container_apps.go:407]: I replaced Required: false with Optional: true, thanks

**Comment by paolosalvatori** [comment]
I successfully tested the change locally by deploying a container app with no `app_port` in the `dapr` block.

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this PR @paolosalvatori!

Could you add a test where `app_port` is omitted so that we can test this behaviour? I also suspect that changes will need to be made to the expand/flatten functions for this block to conditionally set this property if it's been specified.
https://github.com/hashicorp/terraform-provider-azurerm/blob/c67a180a62d05ec12e87d9507beb187fb015dea4/internal/services/containerapps/helpers/container_apps.go#L426-L462

**Comment by paolosalvatori** [comment]
Hi @stephybun, yesterday I successfully tested the change locally without the need to make any change in the `ExpandContainerAppDapr` or the `FlattenContainerAppDapr` functions with two container apps, both using Dapr, where:

- the first explicitly set a value for `app_port` property in the `Dapr` block
- the second one did not specify any value for the `app_port` property in the `Dapr` block

The `app_port` is an integer, optional property, just like the `failure_count_threshold` that does not require any special code in the `expandContainerAppStartupProbe` and `flattenContainerAppStartupProbe` functions.

In my last commit, I added a new test function called `completeWithNoAppPort` to the `container_app_resource_test.go` file. Please check and let me know if I need to make other changes.

**Review by katbyte — APPROVED** [review]
Thanks @paolosalvatori - LGTM 🦕

File: internal/services/containerapps/container_app_resource_test.go [hunk 2]
```diff
@@ -472,6 +472,101 @@ resource "azurerm_container_app" "test" {
 `, r.templatePlusExtras(data), data.RandomInteger, revisionSuffix)
 }
 
+func (r ContainerAppResource) completeWithNoAppPort(data acceptance.TestData, revisionSuffix string) string {
```
Comment (katbyte) [inline internal/services/containerapps/container_app_resource_test.go:475]: I missed this: you'll want to add a function using this test config, ie
```
func TestAccContainerAppResource_completeWithSidecar(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_container_app", "test")
	r := ContainerAppResource{}
	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.completeWithSidecar(data, "rev1"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
	})
}
````

**Review by katbyte — APPROVED** [review]
THanks @paolosalvatori ! LGTM 🥡

**Comment by paolosalvatori** [comment]
Hi @stephybun could you please review and approve this PR? Thanks 😸

**Comment by paolosalvatori** [comment]
When approved and merged, will this feature be included in the next release? If yes, what is the ETA for the next release? 😄

**Comment by ilmax** [comment]
@paolosalvatori usually azurerm gets released every Friday

**Comment by jackofallops** [comment]
> Hi @stephybun, yesterday I successfully tested the change locally without the need to make any change in the `ExpandContainerAppDapr` or the `FlattenContainerAppDapr` functions with two container apps, both using Dapr, where:
> 
> * the first explicitly set a value for `app_port` property in the `Dapr` block
> * the second one did not specify any value for the `app_port` property in the `Dapr` block
> 
> The `app_port` is an integer, optional property, just like the `failure_count_threshold` that does not require any special code in the `expandContainerAppStartupProbe` and `flattenContainerAppStartupProbe` functions.
> 
> In my last commit, I added a new test function called `completeWithNoAppPort` to the `container_app_resource_test.go` file. Please check and let me know if I need to make other changes.

Thanks for the work so far @paolosalvatori - The mention around checking the expand/flatten is to ensure that the zero'd values are actually honoured in cases where they are removed. Can you add a test case that does this? (You should be able to simply update the `completeUpdate` config to remove the `app_port` line from the config there to satisfy this testing requirement).

wrt release - we release weekly on a Thursday, and provided a PR is tested and merged before that, it will ship in that release. We do ask that contributors don't ping on these tasks as the notifications can be distracting and can actually slow the review process down, so I'm going to mark the above comments as off-topic.

**Comment by jackofallops** [comment]
> please let me know if this enough or if you meant something else.

Hi @paolosalvatori - Apologies, I should have been more clear. The test you already added demonstrates that the `Create` will work without the `app_port` property, however, the expand and flatten functions are also used in the `Update`, so we need to ensure test coverage is also present for that path too. Some API's require specific changes to remove values from properties, so to satisfy that Terraform will in fact remove the `app_port` value from the resource successfully by removing the property we need a test that performs this.  I believe the suggestion I made in the above comment, i.e. removing the `app_port` value from `completeUpdate` should provide this coverage and confidence that no other code changes are required (assuming the tests pass, ofc)

Hope this makes things more clear?
Thanks!

**Comment by paolosalvatori** [comment]
Hi @jackofallops / @katbyte 

I added the following functions. Please let me know if this is what you have in mind. If not, please suggest a code change. I'm deploying the complete configuration as a baseline, then I remove the dapr.app_port and finally check that its value is empty.

```
func TestAccContainerAppResource_removeDaprAppPort(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_container_app", "test")
	r := ContainerAppResource{}

	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.complete(data, "rev1"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
		{
			Config: r.completeUpdate_withNoDaprAppPort(data, "rev2"),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).Key("dapr.app_port").IsEmpty(),
			),
		},
		data.ImportStep(),
	})
}

func (r ContainerAppResource) completeUpdate_withNoDaprAppPort(data acceptance.TestData, revisionSuffix string) string {
	return fmt.Sprintf(`
%s

resource "azurerm_container_app" "test" {
  name                         = "acctest-capp-%[2]d"
  resource_group_name          = azurerm_resource_group.test.name
  container_app_environment_id = azurerm_container_app_environment.test.id
  revision_mode                = "Multiple"

  template {
    container {
      name  = "acctest-cont-%[2]d"
      image = "jackofallops/azure-containerapps-python-acctest:v0.0.1"

      cpu    = 0.5
      memory = "1Gi"

      readiness_probe {
        transport               = "HTTP"
        port                    = 5000
        path                    = "/uptime"
        timeout                 = 2
        failure_count_threshold = 1
        success_count_threshold = 1

        header {
          name  = "Cache-Control"
          value = "no-cache"
        }
      }

      liveness_probe {
        transport = "HTTP"
        port      = 5000
        path      = "/health"

        header {
          name  = "Cache-Control"
          value = "no-cache"
        }

        initial_delay           = 5
        timeout                 = 2
        failure_count_threshold = 3
      }

      startup_probe {
        transport               = "TCP"
        port                    = 5000
        timeout                 = 5
        failure_count_threshold = 1
      }

      volume_mounts {
        name = azurerm_container_app_environment_storage.test.name
        path = "/tmp/testdata"
      }
    }

    volume {
      name         = azurerm_container_app_environment_storage.test.name
      storage_type = "AzureFile"
      storage_name = azurerm_container_app_environment_storage.test.name
    }

    min_replicas = 1
    max_replicas = 4

    revision_suffix = "%[3]s"
  }

  ingress {
    allow_insecure_connections = true
    external_enabled           = true
    target_port                = 5000
    transport                  = "auto"

    traffic_weight {
      latest_revision = true
      percentage      = 20
    }

    traffic_weight {
      revision_suffix = "rev1"
      percentage      = 80
    }
  }

  registry {
    server               = azurerm_container_registry.test.login_server
    username             = azurerm_container_registry.test.admin_username
    password_secret_name = "registry-password"
  }

  secret {
    name  = "registry-password"
    value = azurerm_container_registry.test.admin_password
  }

  secret {
    name  = "rick"
    value = "morty"
  }

  dapr {
    app_id       = "acctest-cont-%[2]d"
  }

  tags = {
    foo     = "Bar"
    accTest = "1"
  }
}
`, r.templatePlusExtras(data), data.RandomInteger, revisionSuffix)
}
```

**Comment by paolosalvatori** [comment]
Hi  @jackofallops / @katbyte can you please review my change and let me know if it works? I couldn't test the latest test function.

**Comment by paolosalvatori** [comment]
Hi @stephybun can you please look at this PR, so we merge the fix? Thanks 😄

File: internal/services/containerapps/container_app_resource_test.go [hunk 3]
```diff
@@ -195,6 +210,28 @@ func TestAccContainerAppResource_completeUpdate(t *testing.T) {
...
+			Config: r.complete(data, "rev1"),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+		{
+			Config: r.completeUpdate_withNoDaprAppPort(data, "rev2"),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).Key("dapr.app_port").IsEmpty(),
```
Comment (jackofallops) [inline internal/services/containerapps/container_app_resource_test.go:228]: I think the path here would need to be 
```suggestion
				check.That(data.ResourceName).Key("dapr.0.app_port").IsEmpty(),
```
for the test to evaluate correctly.

**Comment by paolosalvatori** [comment]
@stephybun I committed your suggestion, thanks!

**Comment by katbyte** [comment]
@paolosalvatori - looks like the test check is failing with
```
------- Stdout: -------
=== RUN   TestAccContainerAppResource_removeDaprAppPort
=== PAUSE TestAccContainerAppResource_removeDaprAppPort
=== CONT  TestAccContainerAppResource_removeDaprAppPort
    testcase.go:110: Step 3/4 error: Check failed: Check 1/1 error: azurerm_container_app.test: Attribute 'dapr.0.app_port' expected "", got "0"
--- FAIL: TestAccContainerAppResource_removeDaprAppPort (1017.08s)
FAIL
```

should just need to update the test check?

**Comment by paolosalvatori** [comment]
@katbyte @jackofallops can you please check the PR now? It looks to me that now all the tests succeed

**Review by katbyte — APPROVED** [review]
thanks @paolosalvatori ! tests pass now LGTM 🚟

**Comment by paolosalvatori** [comment]
Thanks @katbyte / @jackofallops for the collaboration.

**Pull request merged**